pgmpy/models/FactorGraph.py

```diff
@@ -200,6 +193,9 @@ def check_model(self):
             raise ValueError('Factors not associated with all the factor nodes.')
 
         cardinalities = self.get_cardinality()
+        if len(self.get_variable_nodes()) != len(cardinalities):
+            raise ValueError('Factors for all the variables not defined')
+
```

Comment:
@ankurankan IMO we need not perform this check(earlier this check was in get_cardinality method) in this method at all .
The following check of check_model will be sufficient 
```
if not all(isinstance(factor_node, DiscreteFactor) for factor_node in factor_nodes):
            raise ValueError('Factors not associated for all the random variables')  
```

Reply:
For example,
G = FactorGraph()
G.add_nodes_from(['a', 'b', 'c'])
phi1 = DiscreteFactor(['a', 'b'], [2, 2], np.random.rand(4))
G.add_nodes_from([phi1])
G.add_factors(phi1)
G.add_edges_from([('a', phi1), ('b', phi1)])
G.check_model()

OUTPUT-
ValueError: Factors not associated for all the random variables

Reply:
@ankurankan I just though of discussing it with you so that i can make the final PR accordingly.

Reply:
@lohani2280 I think you are right. Just the problem is that `get_variable_nodes` calls `check_model`, so you will be stuck in an infinite loop. But I am sure there should be a way around this.

Reply:
@ankurankan I tried to overcome this infinite loop problem by using `variable_nodes` variable instead of `get_variable_nodes` method. Please have a look at my updated PR and review it.